### DISCUSSION TITLE
Is raycasting possible?

### DISCUSSION THREAD

**Discussion by Stingaaa** [cogentcore/core#1352] [Q&A] [answered]
I am currently trying to get a point in the xyz scene based on where i clicked on the screen.
I now have a working function which casts a correct ray but I am currently facing a problem with the intersection.
It seems like the object that gets intersected is the same as the "Selection Box" when I click an object.
Therefore I wanted to ask if it is possible to get the intersection exactly where the object is, if so, could I get an explanation/ example so I can implement this?

**Comment by rcoreilly** [comment, marked as answer]
Here's an example for a plane, which I use for my neural network visualization, here: https://github.com/emer/emergent/blob/main/netview/events.go -- see `LayerUnitAtPoint`, which takes an image point and returns the x,y coordinates in the plane.  The core code is here:

```Go
ray := lo.RayPick(pos)
// layer is in XZ plane with norm pointing up in Y axis
// offset is 0 in local coordinates
plane := math32.Plane{Norm: math32.Vec3(0, 1, 0), Off: 0}
pt, ok := ray.IntersectPlane(plane)
```

`lo` is an `xyz.Solid`

Basically, the `math32.Ray` object has all the methods you might need to find more specific intersections.

**Reply by Stingaaa** [reply]
Thank you very much for the reply.
I wasn't able to look at it today, but I will try it out tomorrow.
I will mark this as the answer for now, as I believe at a first glance that this will help me solve the remains of my problem.

**Reply by Stingaaa** [reply]
Hello @rcoreilly,
I have been able to get further thanks to your help but now I am facing another problem with raycasting.
I want to get the exact coordinates in the 3-dimensional space and whilst I am able to correctly cast the ray and extract the intersection points, I get multiple intersection points and I do not know how to get the correct one.
It seems to me like the invisible bounding boxes when using the "Selection" mode also count as intersections.
If this is the case, is there a way to disable the ability of those bounding boxes to count as a intersection or is there another function/ solution I am not able to come up with/ find.

**Reply by rcoreilly** [reply]
Think of the bounding boxes as a rough first pass for finding the potentially relevant objects. Then you have to use whatever logic is appropriate for your situation to sort through those objects -- e.g., closest distance or something like that. Depending on your scene etc there may not be a unique target so you just have to figure out what works best.

**Reply by rcoreilly** [reply]
BTW, many efficient algorithms in the 3D world involve this kind of multi-scale process of first narrowing down a set of possible objects to then process further, e.g., for collision analysis.

**Reply by Stingaaa** [reply]
FIrst of, thank you for taking the time and explaining the concept/ common practice to me.
After your explanation it seems obvious but I just didn't see this possibilty/ solution.
Once again, thank you very much, I really appreciate your help and guidance.
